### ISSUE TITLE
Can't successfully install sgdq17-layouts -- get an error when at serialport install

### ISSUE THREAD

**Issue by otisbison** [GamesDoneQuick/sgdq17-layouts#6]
Running on Windows 10. Installed Node.js and NodeCG in just the last few days so I have the latest versions.

I wanted to see a working bundle, but I don't know how to get around this. It looks like it's having trouble finding a necessary package: "serialport@4.0.7"

```
C:\Users\otisbison\Documents\Node Projects\nodecg>nodecg install "GamesDoneQuick/sgdq17-layouts"
Fetching sgdq17-layouts release list... done!
Installing sgdq17-layouts... done!
Installing npm dependencies (dev: false)... failed!
Error: Command failed: npm install --production
node-pre-gyp info it worked if it ends with ok
node-pre-gyp verb cli [ 'C:\\Program Files\\nodejs\\node.exe',
node-pre-gyp verb cli   'C:\\Users\\otisbison\\Documents\\Node Projects\\nodecg\\bundles\\sgdq17-layouts\\node_modules\\serialport\\node_modules\\node-pre-gyp\\bin\\node-pre-gyp',
node-pre-gyp verb cli   'install',
node-pre-gyp verb cli   '--fallback-to-build' ]
node-pre-gyp info using node-pre-gyp@0.6.32
node-pre-gyp info using node@9.3.0 | win32 | x64
node-pre-gyp verb command install []
node-pre-gyp info check checked for "C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport\build\Release\serialport.node" (not found)
node-pre-gyp http GET https://github.com/EmergingTechnologyAdvisors/node-serialport/releases/download/4.0.7/serialport-v4.0.7-node-v59-win32-x64.tar.gz
node-pre-gyp http 404 https://github.com/EmergingTechnologyAdvisors/node-serialport/releases/download/4.0.7/serialport-v4.0.7-node-v59-win32-x64.tar.gz
node-pre-gyp ERR! Tried to download(404): https://github.com/EmergingTechnologyAdvisors/node-serialport/releases/download/4.0.7/serialport-v4.0.7-node-v59-win32-x64.tar.gz
node-pre-gyp ERR! Pre-built binaries not found for serialport@4.0.7 and node@9.3.0 (node-v59 ABI) (falling back to source compile with node-gyp)
node-pre-gyp http 404 status code downloading tarball https://github.com/EmergingTechnologyAdvisors/node-serialport/releases/download/4.0.7/serialport-v4.0.7-node-v59-win32-x64.tar.gz
node-pre-gyp verb command build [ 'rebuild' ]
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "C:\Users\otisbison\AppData\Local\Programs\Python\Python36-32\python.EXE", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (C:\Users\otisbison\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\configure.js:483:19)
gyp ERR! stack     at PythonFinder.<anonymous> (C:\Users\otisbison\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\lib\configure.js:508:16)
gyp ERR! stack     at C:\Users\otisbison\AppData\Roaming\npm\node_modules\npm\node_modules\graceful-fs\polyfills.js:284:29
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:166:21)
gyp ERR! System Windows_NT 10.0.15063
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\otisbison\\AppData\\Roaming\\npm\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "configure" "--fallback-to-build" "--module=C:\\Users\\otisbison\\Documents\\Node Projects\\nodecg\\bundles\\sgdq17-layouts\\node_modules\\serialport\\build\\Release\\serialport.node" "--module_name=serialport" "--module_path=C:\\Users\\otisbison\\Documents\\Node Projects\\nodecg\\bundles\\sgdq17-layouts\\node_modules\\serialport\\build\\Release"
gyp ERR! cwd C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport
gyp ERR! node -v v9.3.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok
node-pre-gyp ERR! build error
node-pre-gyp ERR! stack Error: Failed to execute 'C:\Program Files\nodejs\node.exe C:\Users\otisbison\AppData\Roaming\npm\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js configure --fallback-to-build --module=C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport\build\Release\serialport.node --module_name=serialport --module_path=C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport\build\Release' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport\node_modules\node-pre-gyp\lib\util\compile.js:83:29)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:159:13)
node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:943:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:220:5)
node-pre-gyp ERR! System Windows_NT 10.0.15063
node-pre-gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\otisbison\\Documents\\Node Projects\\nodecg\\bundles\\sgdq17-layouts\\node_modules\\serialport\\node_modules\\node-pre-gyp\\bin\\node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport
node-pre-gyp ERR! node -v v9.3.0
node-pre-gyp ERR! node-pre-gyp -v v0.6.32
node-pre-gyp ERR! not ok
npm WARN rollback Rolling back readable-stream@2.2.2 failed (this is probably harmless): EPERM: operation not permitted, scandir 'C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport\node_modules\node-pre-gyp\node_modules\npmlog\node_modules\are-we-there-yet\node_modules'
npm WARN rollback Rolling back hawk@3.1.3 failed (this is probably harmless): EPERM: operation not permitted, lstat 'C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport\node_modules\node-pre-gyp\node_modules\request\node_modules'
npm WARN rollback Rolling back har-validator@2.0.6 failed (this is probably harmless): EPERM: operation not permitted, lstat 'C:\Users\otisbison\Documents\Node Projects\nodecg\bundles\sgdq17-layouts\node_modules\serialport\node_modules\node-pre-gyp\node_modules\request\node_modules'
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! serialport@4.0.7 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the serialport@4.0.7 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\otisbison\AppData\Roaming\npm-cache\_logs\2017-12-28T15_18_39_020Z-debug.log

    at checkExecSyncError (child_process.js:574:11)
    at execSync (child_process.js:611:13)
    at module.exports (C:\Users\otisbison\AppData\Roaming\npm\node_modules\nodecg-cli\lib\install-bundle-deps.js:30:4)
    at Command.action (C:\Users\otisbison\AppData\Roaming\npm\node_modules\nodecg-cli\commands\install.js:110:2)
    at Command.listener (C:\Users\otisbison\AppData\Roaming\npm\node_modules\nodecg-cli\node_modules\commander\index.js:315:8)
    at Command.emit (events.js:159:13)
    at Command.parseArgs (C:\Users\otisbison\AppData\Roaming\npm\node_modules\nodecg-cli\node_modules\commander\index.js:653:12)
    at Command.parse (C:\Users\otisbison\AppData\Roaming\npm\node_modules\nodecg-cli\node_modules\commander\index.js:475:21)    at Object.<anonymous> (C:\Users\otisbison\AppData\Roaming\npm\node_modules\nodecg-cli\index.js:43:9)
    at Module._compile (module.js:660:30)

C:\Users\otisbison\Documents\Node Projects\nodecg>
```

`

[2017-12-28T15_18_39_020Z-debug.log](https://github.com/GamesDoneQuick/sgdq17-layouts/files/1591472/2017-12-28T15_18_39_020Z-debug.log)

**Comment by Lange** [comment]
Did you follow the [installation instructions](https://github.com/GamesDoneQuick/sgdq17-layouts#installation) in full? They mention that you need to install a compiler toolchain and `npm install` this bundle's dependencies manually. NodeCG will fail to automatically install them for you.

**Comment by otisbison** [comment]
I guess I’m confused about the instructions then..

I started at 1. And took that to mean do : “nodecg install “GamesDoneQuick/sgdq17-layouts” and assumed the build tools came after.

I see now that it says manual npm install required..

I went ahead and installed the build tools.  I did step 4 and it gave me an error.

So is Step 1 actually:

 npm install “GamesDoneQuick/sgdq17-layouts” ?

I tried that and get the same error about serial port not being found.  I looked on the github for EmergingTechnology[...] and the package isn’t there, so maybe there is larger problem here?

**Comment by Lange** [comment]
When it says "manual npm install", it means do this:
```bash
cd nodecg/bundles/sgdq17-layouts
npm install
```

As for the `serialport` dependency, yes it is indeed listed in this bundle's [`package.json`, in line #31](https://github.com/GamesDoneQuick/sgdq17-layouts/blob/master/package.json#L31). The issue you are encountering is that this dependency is not installing correctly. This is common, because correctly installing this dependency often requires building it from source, which requires having a working compiler toolchain set up.

**Comment by Lange** [comment]
Looking more closely at your error log, it says this:
>gyp ERR! stack Error: Can't find Python executable

This is probably one of the main sources of your issues. `windows-build-tools` should have installed Python 2.7 for you and added it to your PATH, but maybe it didn't. You might need to do that manually.

**Comment by Lange** [comment]
Also, you need to close and re-open your terminal (command prompt) after changing things like your PATH.

**Comment by otisbison** [comment]
I seem to have worked it out... the windows build tools log said it successfully installed but I don't believe it actually did. I tried directly installing serialport and finally noticed that it said it couldn't find a "CL.EXE" at the top. This error wasn't showing when I tried installing sgdq17-layouts.

I'm assuming CL.EXE would have been included with the build tools.

I already had Visual Studio 2015 installed and I think it was blocking the build tools. From within VS 2015 I installed the c++ tools and then everything completely properly.
